src/xref2/tools.ml

```diff
@@ -981,88 +1057,101 @@ and reresolve_module : Env.t -> Cpath.Resolved.module_ -> Cpath.Resolved.module_
-                          Odoc_model.Names.ModuleName.make_std n ),
-                      false )
-                in
-                match resolve p' with None -> resolve fallback | x -> x)
-            | _ -> resolve fallback))
-    | _ -> None
+  let resolve env p =
+    resolve_module env ~mark_substituted:false ~add_canonical:false p
+    >>= fun (p, m) -> Ok (strip_alias p, m)
   in
-  match get_cpath p2 with Some p -> `Resolved p | None -> p2
-
-and handle_canonical_module_type env p2 =
-  let resolve p =
-    match
-      resolve_module_type ~mark_substituted:true ~add_canonical:false env p
-    with
-    | Ok (p, _) -> `Resolved p
-    | Error _ -> p2
+  let lang_of cpath =
+    (Lang_of.(Path.resolved_module (empty ()) cpath)
+      :> Odoc_model.Paths.Path.Resolved.t)
   in
-  match p2 with
-  | `Dot (p, n) -> (
-      match handle_canonical_module env p with
-      | `Resolved r as p' -> (
-          let fallback = `Dot (p', n) in
-          match r with
-          | `Identifier pid -> (
-              let p' =
-                `Identifier
-                  ( `ModuleType
-                      ( (pid :> Odoc_model.Paths.Identifier.Signature.t),
-                        Odoc_model.Names.ModuleTypeName.make_std n ),
-                    false )
-              in
-              match resolve p' with
-              | `Resolved _ as x -> x
-              | _ -> resolve fallback)
-          | _ -> resolve fallback)
-      | _ -> p2)
-  | _ -> p2
-
-and handle_canonical_type env p2 =
-  let resolve p =
-    match resolve_type ~add_canonical:false env p with
-    | Ok (p, _) -> `Resolved p
-    | Error _ -> p2
+  match canonical_helper env resolve lang_of c_mod_poss p2 with
+  | None -> p2
+  | Some (rp, m) ->
+      let m = Component.Delayed.get m in
+      (* Need to check if the module we're going to link to has been expanded.
+         ModuleTypes are always expanded if possible, but Aliases are only expanded
+         if they're an alias to a hidden module or if they're self canonical.
+
+         Checking if a module is self canonical is a bit tricky, since this function
+         is itself part of the process of resolving any canonical reference. Hence
+         what we do here is to look through alias chains looking for one that's marked
+         with the same _unresolved_ canonical path that we're currently trying to resolve.
+
+         This is particularly important because some modules don't know they're canonical!
+         For example the module Caml in base, which is marked as the canonical path for
+         all references to the standard library in the file [import0.ml], but is itself just
+         defined by including [Stdlib].
+
+         If a module doesn't know it's canonical, it will fail the self-canonical check, and
+         therefore not necessarily be expanded. If this happens, we call [process_module_path]
+         to stick the [`Alias] constructor back on so we'll link to the correct place. *)
```

Comment:
Shouldn't this be done by the caller ? Because like this, it'll still be wrapped in a `` `Canonical `` and that makes this function part of the complicated recursive functions.

Reply:
I'd prefer to leave it here so we don't spread the logic between two functions. It'd be better if we have one function to test for self-canonical modules that we can call from here and also from `link.ml`, but there's a bit of refactoring required for that.

Reply:
I mean the resolving could fail and the function that handles the `` `Canonical `` case can start again by calling `reresolve_module` and at the same time removing the `` `Canonical `` constructor.

Reply:
Ah, no, the canonical is still valid, it's just that it's going to point to the aliased module rather than the module itself, as the module itself won't be expanded because it doesn't know it's canonical :-)

Reply:
I see.